### PR TITLE
Add signup theme purchase test

### LINKED ISSUE #843: Add test for signup & purchase flow from /themes
As reported in https://github.com/Automattic/wp-calypso/issues/20717, the theme purchase flow from the logged out theme showcase at https://wordpress.com/themes recently broke.

Basic flow:

- Start at logged-out premium theme page, e.g. https://wordpress.com/theme/small-business
- Select "Pick this design"
- Continue through signup
- Confirm that you arrive at checkout with the selected theme in the cart

### PR DESCRIPTION
**Pull request by brbrr**
Fixes #843

### CODE REVIEW

File: lib/components/secure-payment-component.js [hunk 1]
```diff
@@ -41,4 +41,11 @@ export default class SecurePaymentComponent extends BaseContainer {
 		this.driver.wait( until.elementLocated( selector ), this.explicitWaitMS, 'Could not locate the business plan price element' );
 		return this.driver.findElement( selector ).getText();
 	}
+
+	getProductsNames() {
+		const selector = By.css( '.product-name' );
+		return this.driver.findElements( selector ).then( function( products ) {
+			return promise.all( products.map( e => e.getText() ) );
```
Comment (alisterscott) [inline lib/components/secure-payment-component.js]: nice one 👍

File: lib/pages/themes-page.js [hunk 2]
```diff
@@ -1,15 +1,28 @@
...
+		if ( visit === true ) {
+			url = config.get( 'calypsoBaseURL' ) + '/themes';
+		}
+		super( driver, by.css( '.theme__active-focus' ), visit, url );
 		this.waitUntilThemesLoaded();
 	}
 
 	showOnlyFreeThemes() {
-		driverHelper.clickWhenClickable( this.driver, by.css( 'a[data-e2e-value="free"]' ) );
+		return this.showOnlyThemesType( 'free' )
```
Comment (alisterscott) [inline lib/pages/themes-page.js]: needs semicolon

File: lib/pages/themes-page.js [hunk 3]
```diff
@@ -1,15 +1,28 @@
...
+	}
+
+	showOnlyPremiumThemes() {
+		return this.showOnlyThemesType( 'premium' )
```
Comment (alisterscott) [inline lib/pages/themes-page.js]: semicolon

File: lib/pages/themes-page.js [hunk 4]
```diff
@@ -77,7 +90,7 @@ export default class ThemesPage extends BaseContainer {
 		return this.driver.findElement( popOverMenuSelector ).isDisplayed();
 	}
 
-	static _getThemeSelectionXpath(phrase ) {
+	static _getThemeSelectionXpath( phrase ) {
 		return by.xpath( `//h2[text()[contains(.,'${ phrase }')]]/../../../../div[not(contains(concat(' ', @class, ' '), ' is-active '))]//span` );
```
Comment (alisterscott) [inline lib/pages/themes-page.js]: is there anyway to simplify this complex mess (my code I think) - maybe by adding a data attribute to the wp-calypso screen?

File: specs/wp-signup-spec.js [hunk 5]
```diff
@@ -38,6 +38,8 @@ import SideBarComponent from '../lib/components/sidebar-component';
 import * as SlackNotifier from '../lib/slack-notifier';
 
 import EmailClient from '../lib/email-client.js';
+import ThemesPage from '../lib/pages/themes-page';
+import ThemePage from '../lib/pages/theme-page';
```
Comment (alisterscott) [inline specs/wp-signup-spec.js]: The file seems missing, should you use `lib/pages/theme-detail-page.js` and just add `pickThisDesign` to it?

File: specs/wp-signup-spec.js [hunk 6]
```diff
@@ -1173,4 +1175,127 @@ testDescribe( `[${host}] Sign Up  (${screenSize}, ${locale})`, function() {
...
+			test.it( 'Can select premium theme', function() {
+				this.name = ''
+				this.themesPage.showOnlyPremiumThemes()
+				this.themesPage.getFirstThemeName().then( name => this.name = name )
+				return this.themesPage.selectNewTheme();
+			} );
+
+			test.it( 'Can pick theme design', function() {
+				this.themePage = new ThemePage( driver )
+				return this.themePage.pickThisDesign()
```
Comment (alisterscott) [inline specs/wp-signup-spec.js]: This doesn't exist at the moment

File: specs/wp-signup-spec.js [hunk 7]
```diff
@@ -1173,4 +1175,127 @@ testDescribe( `[${host}] Sign Up  (${screenSize}, ${locale})`, function() {
...
+								stepNum++;
+
+								test.it( 'Can then see the secure payment page', function() {
+									this.securePaymentComponent = new SecurePaymentComponent( driver );
+									return this.securePaymentComponent.displayed().then( ( displayed ) => {
+										return assert.equal( displayed, true, 'The secure payment page is not displayed' );
+									} );
+								} );
+
+								test.it( 'Can enter and submit test payment details', function() {
```
Comment (alisterscott) [inline specs/wp-signup-spec.js]: are we actually submitting this? if not, probably best to rename this step

**Comment by brbrr** [comment]
Ouch, looks like Theme switch test become flacky. It passed after 2nd run on CI.
Failed run: https://circleci.com/gh/Automattic/wp-e2e-tests/15946

**Comment by alisterscott** [comment]
> Ouch, looks like Theme switch test become flacky. It passed after 2nd run on CI.

Did you change that test? Or just the pages it uses?

**Comment by brbrr** [comment]
> Did you change that test? Or just the pages it uses?

Only page. But I already found an issue - updated `xpath` was selecting all themes, even active theme. Now it should be resolved

File: lib/pages/themes-page.js [hunk 8]
```diff
@@ -91,8 +91,7 @@ export default class ThemesPage extends BaseContainer {
 	}
 
 	static _getThemeSelectionXpath( phrase ) {
-		// return by.xpath( `//h2[text()[contains(.,'${ phrase }')]]//ancestor::div[contains(@class,'card')]` );
-		//h2[text()[contains(.,'Twenty')]]//ancestor::div[contains(@class,'card') && not(contains(@class,'is-active'))]
-		return by.xpath( `//h2[text()[contains(.,'${ phrase }')]]//ancestor::div[contains(@class,'card') and not(contains(@class,'is-active'))]` );
+		let lowerCasedPhrase = phrase.toLowerCase().replace( ' ', '-' );
+		return by.css( `div[data-e2e-theme*='${ lowerCasedPhrase }']:not(.is-active) span` );
```
Comment (alisterscott) [inline lib/pages/themes-page.js]: 💯

**Pull request merged**